### PR TITLE
973 - concatenate the value + extracted text for both label for attribute a…

### LINKED ISSUE #973: Duplicate labels/ids for Django admin checkboxes
### Current Behavior

The ids and labels for the Django admin checkboxes are being pulled from the text of the anchor element in the next column, which is the model identifier. For Domain Applications when a Started DA has no requested domain yet, that value is "-". 

Consequently, if there are 2 or more started DAs with no requested domains, checkboxes with duplicate ids and labels are detected on the page.

The fix is to concatenate the value + extracted text for both label for attribute and checkbox id attribute.

### Expected Behavior

No duplicate ids or labels on any Django admin page

### Steps to Reproduce

1. Log into django admin
2. Create 2 new DAs
3. Run ANDI on DA list page

### Environment

_No response_

### Additional Context

_No response_

### Issue Links

_No response_

### PR DESCRIPTION
**Pull request by rachidatecs**
…nd checkbox id attribute

## Ticket

Resolves #973 

## Changes

- Concatenate the value + extracted text for both label for attribute and checkbox id attribute

## Context for reviewers

The ids and labels for the Django admin checkboxes are being pulled from the text of the anchor element in the next column, which is the model identifier. For Domain Applications when a Started DA has no requested domain yet, that value is "-".

Consequently, if there are 2 or more started DAs with no requested domains, checkboxes with duplicate ids and labels are detected on the page.

The fix is to concatenate the value + extracted text for both label for attribute and checkbox id attribute.

#### Satisfied acceptance criteria and met development standards

- [x] Met the acceptance criteria, or will meet them in a subsequent PR
- [ ] Created/modified automated tests
- [x] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [ ] Messaged on Slack or in standup to notify the team that a PR is ready for review
- [ ] Changes to “how we do things” are documented in READMEs and or onboarding guide
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Original Developer)

- [ ] All new functions and methods are commented using plain language
- [ ] Did dependency updates in Pipfile also get changed in requirements.txt?
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] Add at least 1 designer as PR reviewer

#### Reviewed, tested, and left feedback about the changes

- [ ] Pulled this branch locally and tested it
- [ ] Reviewed this code and left comments
- [ ] Checked that all code is adequately covered by tests
- [ ] Made it clear which comments need to be addressed before this work is merged
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [ ] All new functions and methods are commented using plain language
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [ ] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

## Screenshots

Screen Shot 2023-09-01 at 5 43 23 PM

### CODE REVIEW

**Comment by therealslimhsiehdy** [comment]
Just to clarify, if a user has two started Domain Applications the screen reader will read the same IDs and labels?

(The code LGTM but just wanted to confirm context)

File: src/registrar/templates/admin/change_list_results.html [hunk 1]
```diff
@@ -60,12 +60,11 @@
...
     {% endif %}
     <tr>
-
         {% with result_value=result.0|extract_value %}
             {% with result_label=result.1|extract_a_text %}
                 <td>
-                    <input type="checkbox" name="_selected_action" value="{{ result_value|default:'value' }}" id="{{ result_label|default:result_value }}" class="action-select">
-                    <label class="usa-sr-only" for="{{ result_label|default:result_value }}">{{ result_label|default:'label' }}</label>
+                    <input type="checkbox" name="_selected_action" value="{{ result_value|default:'value' }}" id="{{ result_value|default:'value' }} {{ result_label|default:'label' }}" class="action-select">
+                    <label class="usa-sr-only" for="{{ result_value|default:'value' }} {{ result_label|default:'label' }}">{{ result_label|default:'label' }}</label>
```
Comment (neilmb) [inline src/registrar/templates/admin/change_list_results.html:67]: **suggestion**: "An id's value must not contain whitespace" https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/id

```suggestion
                    <input type="checkbox" name="_selected_action" value="{{ result_value|default:'value' }}" id="{{ result_value|default:'value' }}-{{ result_label|default:'label' }}" class="action-select">
                    <label class="usa-sr-only" for="{{ result_value|default:'value' }}-{{ result_label|default:'label' }}">{{ result_label|default:'label' }}</label>
```

**Review by neilmb — COMMENTED** [review]
Just a tweak to get rid of whitespace in the `id` attribute.

File: src/registrar/templates/admin/change_list_results.html (hunk 1 above)
Comment (rachidatecs) [reply src/registrar/templates/admin/change_list_results.html:67]: Good catch!

**Comment by therealslimhsiehdy** [comment]
> > Just to clarify, if a user has two started Domain Applications the screen reader will read the same IDs and labels?
> > (The code LGTM but just wanted to confirm context)
> 
> The id and label for each started application will be "-", and therefore duplicates.

Thank you for clarifying! I added two screenshots just for visual reference (feel free to update the PR with them if you'd like!):

On my local without your changes, you can see in the red circles that the IDs are duplicated as "-":
Screenshot 2023-09-05 at 3 31 28 PM

On staging with your newest wonderful changes, you can see in the red circles now that the IDs are different with "81--" and "82--":
Screenshot 2023-09-05 at 3 33 19 PM

Huge impact!

**Review by therealslimhsiehdy — APPROVED** [review]

**Comment by rachidatecs** [comment]
You just added the screenshots to the PR, @therealslimhsiehdy  :)
Thanks for the thorough testing.

Do we care that in case of a non-existing result_label string (the string we're pulling from the anchor text, ie the requested domain), Django defaults to '-'. So the concatenated uniques string is now id + '-' + '-' = id--

It still meets the goal of providing accessible and unique  ids and labels, just ain't super pretty. Maybe follow-up with a small ticket to replace '-' with 'new-application', resulting in strings that look like {id}-new-application?

**Comment by zandercymatics** [comment]
@rachidatecs Not Rebecca but I would personally say that what you said is worth addressing only if in the foreseeable future we need to use those ids for something. Django seems to handle those id changes fine (the 'for' label), so that additional dash doesn't seem to break anything. 

Fixing it would benefit the following: 
a) future development using these ids (would make it slightly easier to figure out / deal with)
- If this is a concern, personally I would say documenting this as a comment would be sufficient. (Ex: will result in an id of 'x--') 

b) html/code cleanliness 
- I don't think this negatively detracts from this as its a small thing. In general, I would argue that code cleanliness should be a priority where many eyes will lie and work collaboratively. If its an area that we (very likely) will not touch again, then imo I don't personally consider it necessary.

**TLDR:** I don't think there's any harm in creating a follow-up ticket for this, but I personally don't think this is something we _need_ to address. It falls into the "nice to have" category.

**Comment by rachidatecs** [comment]
I'm thinking this could be an improvement for screen readers. The 'for' attribute gets read out-loud, no?

I could ask Aaron...

**Comment by zandercymatics** [comment]
@rachidatecs Ohh... Yes that is an excellent point. I did not even think of that. If it does get read aloud then 100% we should get this fixed at a later date

**Comment by neilmb** [comment]
`for` does not get read out loud. It is the link between the label's `id` and the input element that corresponds to the label so that the screenreader can announce the correct label when the input element is selected. IDs ought never be exposed to the user, so as long as they are unique, preventing collisions for situations like this, they can have as many dashes as they want.
